draft-ietf-mls-protocol.md

```diff
@@ -1524,223 +1460,285 @@ The creator of the group constructs an Init message as follows:
 
-The `welcome_info_hash` field contains a hash of the WelcomeInfo
-object sent in a Welcome message to the new member.
+* Update the sender's leaf node by replacing the HPKE public key with the public
+  key in the Update proposal
 
-A group member generates this message by requesting a ClientInitKey
-from the directory for the user to be added, and encoding it into an
-Add message.
+* Blank the intermediate nodes along the path from the sender's leaf to the root
 
-The client joining the group processes Welcome and Add
-messages together as follows:
+### Remove
 
-* Prepare a new GroupContext object based on the Welcome message
-* Process the Add message as an existing member would
+A Remove proposal requests that the client at a specified index in the tree be
+removed from the group.
 
-An existing member receiving a Add message first verifies
-the signature on the message,  then updates its state as follows:
+~~~~~
+struct {
+    uint32 removed;
+} Remove;
+~~~~~
 
-* If the `index` value is equal to the size of the group, increment
-  the size of the group, and extend the tree accordingly
-* Verify the signature on the included ClientInitKey; if the signature
-  verification fails, abort
-* Generate a WelcomeInfo object describing the state prior to the
-  add, and verify that its hash is the same as the value of the
-  `welcome_info_hash` field
-* Update the ratchet tree by setting to blank all nodes in the
-  direct path of the new node
-* Set the leaf node in the tree at position `index` to a new node
-  containing the public key from the ClientInitKey in the Add, as
-  well as the credential under which the ClientInitKey was signed
+A member of the group applies a Remove message by taking the following steps:
 
-The `update_secret` resulting from this change is an all-zero octet
-string of length Hash.length.
+* Replace the leaf node at position `removed` with a blank node
 
-After processing an Add message, the new member SHOULD send an Update
-immediately to update its key. This will help to limit the tree structure
-degrading into subtrees, and thus maintain the protocol's efficiency.
+* Blank the intermediate nodes along the path from the removed leaf to the root
 
-## Update
+## Commit
 
-An Update message is sent by a group member to update its leaf
-secret and key pair.  This operation provides post-compromise security
-with regard to the member's prior leaf private key.
+A Commit message initiates a new epoch for the group, based on a collection of
+proposals.  It instructs group members to update their representation of the
+state of the group by applying the proposals and advancing the key schedule.
+
+A group member that has observed one or more Proposal messages within an epoch
+MUST send a Commit message before sending application data.  This ensures, for
+example, that any members whose removal was proposed during the epoch are
+actually removed before any application information is transmitted.
```

Comment:
Because proposals can't build on top of each other, you have weird cases like:

1. Alice adds Bob to the group (without committing).
2. Carol removes Bob from the group.

It implicitly requires Carol to Commit->Remove because she can't form her Remove proposal without first committing. Whatever API you expose to an application would have to understand this

Reply:
This seems workable to me.  I would expect that the API just wouldn't mark Bob as added until the Add was committed, so Carol wouldn't even be able to form the Remove message until then.

Reply:
Richard is correct here.